tools/deployment_tool/contract.js

```diff
@@ -0,0 +1,323 @@
+        return new Contract(
+            truffleJson.contractName,
+            truffleJson.bytecode,
+            constructorAbi,
+            constructorArgs,
+        );
+    }
+
+    /**
+     * Add a linked dependency to the contract.
+     */
+    addLinkedDependency(linkedContract) {
+        this._ensureState(STATE_NEW);
+        this.linkReplacements.push(linkedContract);
+    }
+
+    /**
+     * Returns the contract's address as detemined by the provided AddressGenerator.
+     *
+     * @param {Object|string} addressGeneratorOrAddress The AddressGenerator used for
+     *                        generating the addresses. Can alternatively be a fixed address.
+     * @returns {string} The contract's address.
+     */
+    setAddress(addressGeneratorOrAddress) {
+        this._ensureStateOneOf([STATE_NEW, STATE_LINKED]);
+        // Return early if the address has previously been set.
+        // This allows for setting a fixed address for specific contracts.
+        if (this.address) {
+            return this.address;
+        }
+
+        if (typeof addressGeneratorOrAddress === 'string') {
+            this.address = addressGeneratorOrAddress;
+            return this.address;
+        }
+
+        const addressGenerator = addressGeneratorOrAddress;
+        if (!addressGenerator) {
+            throw new Error(`addressGenerator not provided when generating address for ${this.contractName}`);
+        }
+
+        const address = addressGenerator.generateAddress();
+        this.address = address;
+        return this.address;
+    }
+
+    /**
+     * Get the address of an instantiated contract.
+     *
+     * @returns {string} The address of the instantiated contract.
+     */
+    getAddress() {
+        this._ensureState(STATE_INSTANTIATED);
+        return this.address;
+    }
+
+    /**
+     * Returns the previously linked bytecode.
+     *
+     * @returns {string} The linked bytecode of the contract.
+     */
+    linkedBytecode() {
+        this._ensureStateOneOf([STATE_LINKED, STATE_INSTANTIATED]);
+        return this.contractBytecode;
+    }
+
+    /**
+     * Instantiate the contract by calculating the data of the contract creation transaction.
+     * This includes encoding the provided constructor arguments.
+     * Freezes the object to prevent any further changes.
+     *
+     * @return {string} The transaction data for contract creation.
+     */
+    instantiate() {
+        this._ensureState(STATE_NEW);
+
+        if (this.constructorABI && !this.constructorArgs) {
+            throw new Error(`Expected constructor arguments for constract ${this.contractName}`);
+        }
+        if (this.constructorArgs && !this.constructorABI) {
+            throw new Error(`Provided arguments for contract ${this.contractName}, but no constructorAbi is set.`);
+        }
+
+        this._linkBytecode();
+
+        let constructorData = this.linkedBytecode();
+        constructorData += this._encodeConstructorArguments();
```

Comment:
Mind adding a little bit of documentation to the above two lines? That would really help me a lot in understanding it.